common/ApplicationConfig.cc.in

```diff
@@ -17,162 +17,611 @@
 
 const std::string Config::CMSTkModLabBasePath("@basepath@");
 
+bool operator<(const ApplicationConfig::FullKey& k1, const ApplicationConfig::FullKey& k2)
+{
+  if (k1.alias == k2.alias) {
+    return k1.key < k2.key;
+  } else {
+    return k1.alias < k2.alias;
+  }
+}
+
+bool operator<(const ApplicationConfig::FullKey& fk, const ApplicationConfig::ShortKey& sk)
+{
+  if (fk.alias == sk.alias) {
+    return fk.key < sk.key;
+  }
+  return fk.alias < sk.alias;
+}
+
+bool operator<(const ApplicationConfig::ShortKey& sk, const ApplicationConfig::FullKey& fk)
+{
+  if (sk.alias == fk.alias) {
+    return sk.key < fk.key;
+  }
+  return sk.alias < fk.alias;
+}
+
 ApplicationConfig* ApplicationConfig::instance_ = NULL;
 
-ApplicationConfig::ApplicationConfig(const std::string filename)
+ApplicationConfig::ApplicationConfig(QObject *parent)
+ : QObject(parent)
 {
-  ApplicationConfigReader reader(filename);
-  reader.fill(keyvalueMap_);
+
+}
+
+ApplicationConfig* ApplicationConfig::instance(QObject *parent)
+{
+  if (instance_==NULL) {
+    instance_ = new ApplicationConfig(parent);
+  }
+  return instance_;
 }
 
-ApplicationConfig* ApplicationConfig::instance(const std::string filename)
+ApplicationConfig* ApplicationConfig::instance(const std::string filename,
+  const std::string alias, QObject *parent)
 {
   if (instance_==NULL) {
+    instance_ = new ApplicationConfig(parent);
     if (!filename.empty()) {
-      instance_ = new ApplicationConfig(filename);
+      instance_->append(filename, alias);
     }
   }
   return instance_;
 }
 
-void ApplicationConfig::safe(const std::string filename)
+void ApplicationConfig::append(const std::string& filename, const std::string& alias)
+{
+  ApplicationConfigReader reader(filename, alias);
+  reader.fill(keyvalueMap_);
+  
+  emit structureChanged();
+  emit valueChanged();
+}
+
+void ApplicationConfig::save()
+{
+  std::map<std::string,std::string> filemap;
+  for (const auto & kv : keyvalueMap_) {
+  	if (kv.first.alias=="*") continue;
+  	filemap[kv.first.filename] = kv.first.alias;
+  }
+
+  for (const auto & kv : filemap) {
+  	ApplicationConfigWriter writer(kv.first, kv.second);
+    writer.write(keyvalueMap_);
+  }
+}
+
+void ApplicationConfig::saveAs(const std::string& filename, const std::string& alias)
 {
-  ApplicationConfigWriter writer(filename);
-  writer.write(keyvalueMap_);
+  storage_t tempKeyValueMap;
+  
+  FullKey fk;
+  fk.filename = filename;
+  fk.alias = alias;
+  
+  for (const auto & kv : keyvalueMap_) {
+  	if (kv.first.alias!="*") continue;
+
+  	fk.key = kv.first.key;
+  	tempKeyValueMap[fk] = kv.second;
+  }
+
+  ApplicationConfigWriter writer(filename, alias);
+  writer.write(tempKeyValueMap);
+  
+  for (const auto & kv : tempKeyValueMap) {
+  	if (kv.first.alias!="*") continue;
+
+	fk.filename = "*";
+  	fk.alias = "*";
+  	fk.key = kv.first.key;
+  	tempKeyValueMap[fk] = kv.second;
+  }
+  
+  keyvalueMap_ = tempKeyValueMap;
```

Comment:
Here we overwrite the configuration with only `*` alias/filenames, so all alias/filename history is gone. Why is this done here? Naively I wouldn't expect a `save` method to change the configuration list.

Reply:
How do we want to do this in case of saveAs? Option one would be that saveAs simply writes the key/value pairs for a given alias to a new file leaving the FullKeys in the storage map untouched. Option two would be that a saveAs changes the filename for a given alias in the FullKeys to the new filename.

Reply:
Huh, good point. Currently the filename is not really used for searching, but it is used for `Save`, so it's not irrelevant. I agree that the `filename` could be updated. I however don't see a need for changing the alias, especially since it can't be done selectively (e.g. "please change the alias A to alias B, but don't touch alias C" or similar).

Reply:
Meaning, I'd be in favour of option two.

Reply:
Fixed in latest commits.